File: backend/src/modules/public/public.service.ts
import { Post, Category, Tag } from '@database/index';
import { AppError } from '@utils/AppError';
import { parsePagination } from '@utils/pagination';
import { POST_STATUS, POST_VISIBILITY } from '@config/constants';
import { Op, WhereOptions } from 'sequelize';

const PUBLIC_WHERE = {
  status: POST_STATUS.PUBLISHED,
  visibility: POST_VISIBILITY.PUBLIC,
};

const POST_INCLUDE = [
  { model: Tag, as: 'tags', through: { attributes: [] }, attributes: ['id', 'name', 'slug'] },
  { model: Category, as: 'category', attributes: ['id', 'name', 'slug'] },
];

export async function listPublicPosts(filters: {
  page?: string;
  limit?: string;
  search?: string;
  category?: string;
  tag?: string;
  sort?: string;
  order?: 'asc' | 'desc';
}) {
  const { page, limit, offset, sort, order } = parsePagination(filters);

  const where: WhereOptions<Post> = { ...PUBLIC_WHERE };

  if (filters.search) {
    where[Op.or as never] = [
      { title: { [Op.like]: `%${filters.search}%` } },
      { excerpt: { [Op.like]: `%${filters.search}%` } },
    ];
  }

  if (filters.category) {
    const category = await Category.findOne({ where: { slug: filters.category } });
    if (category) {
      where.categoryId = category.id;
    }
  }

  const tagInclude = filters.tag
    ? { model: Tag, as: 'tags', where: { slug: filters.tag }, required: true, through: { attributes: [] } }
    : { model: Tag, as: 'tags', through: { attributes: [] }, attributes: ['id', 'name', 'slug'] };

  const { rows, count } = await Post.findAndCountAll({
    where,
    include: [
      tagInclude,
      { model: Category, as: 'category', attributes: ['id', 'name', 'slug'] },
    ],
    limit,
    offset,
    order: [[sort, order]],
    distinct: true,
  });

  return {
    posts: rows,
    meta: { page, limit, total: count, totalPages: Math.ceil(count / limit) },
  };
}

export async function getPublicPostBySlug(slug: string) {
  const post = await Post.findOne({
    where: { slug, ...PUBLIC_WHERE },
    include: POST_INCLUDE,
  });
  if (!post) {
    throw AppError.notFound('Post not found');
  }
  return post;
}

export async function listPublicCategories(filters: { page?: string; limit?: string; search?: string; sort?: string; order?: 'asc' | 'desc' }) {
  const { page, limit, offset, sort, order } = parsePagination(filters);

  const where: WhereOptions<Category> = { status: 'active' };
  if (filters.search) {
    where[Op.or as never] = [{ name: { [Op.like]: `%${filters.search}%` } }];
  }

  const { rows, count } = await Category.findAndCountAll({
    where,
    limit,
    offset,
    order: [[sort, order]],
    distinct: true,
  });

  return {
    categories: rows,
    meta: { page, limit, total: count, totalPages: Math.ceil(count / limit) },
  };
}

export async function getPublicCategoryBySlug(slug: string) {
  const category = await Category.findOne({ where: { slug, status: 'active' } });
  if (!category) {
    throw AppError.notFound('Category not found');
  }
  return category;
}

export async function listPublicTags(filters: { page?: string; limit?: string; search?: string; sort?: string; order?: 'asc' | 'desc' }) {
  const { page, limit, offset, sort, order } = parsePagination(filters);

  const where: WhereOptions<Tag> = {};
  if (filters.search) {
    where[Op.or as never] = [{ name: { [Op.like]: `%${filters.search}%` } }];
  }

  const { rows, count } = await Tag.findAndCountAll({
    where,
    limit,
    offset,
    order: [[sort, order]],
    distinct: true,
  });

  return {
    tags: rows,
    meta: { page, limit, total: count, totalPages: Math.ceil(count / limit) },
  };
}

export async function getPublicTagBySlug(slug: string) {
  const tag = await Tag.findOne({ where: { slug } });
  if (!tag) {
    throw AppError.notFound('Tag not found');
  }
  return tag;
}

export async function searchPublic(query: string, filters: { page?: string; limit?: string }) {
  const { page, limit, offset } = parsePagination(filters);

  const where: WhereOptions<Post> = {
    ...PUBLIC_WHERE,
    [Op.or as never]: [
      { title: { [Op.like]: `%${query}%` } },
      { excerpt: { [Op.like]: `%${query}%` } },
      { content: { [Op.like]: `%${query}%` } },
    ],
  };

  const { rows, count } = await Post.findAndCountAll({
    where,
    limit,
    offset,
    include: POST_INCLUDE,
    order: [['published_at', 'DESC']],
    distinct: true,
  });

  return {
    posts: rows,
    meta: { page, limit, total: count, totalPages: Math.ceil(count / limit) },
  };
}

File: backend/app/services/public_service.py
from app.config import Const
from app.core.errors import AppError
from app.core.pagination import order_by_clause, parse_pagination
from app.core.sanitize import sanitize_content
from app.models import Category, Post, PostTag, Tag


def db_public_condition():
    return Post.status == Const.POST_PUBLISHED, Post.visibility == Const.POST_VISIBILITY_PUBLIC


def _to_public_post(post):
    if post.content:
        post.content = sanitize_content(post.content)
    return post


def list_public_posts(db, filters):
    page, limit, offset, sort, order = parse_pagination(filters)

    q = db.query(Post).filter(Post.deletedAt.is_(None), *db_public_condition())

    if filters.get('search'):
        term = '%{}%'.format(filters['search'])
        q = q.filter(db.or_(Post.title.like(term), Post.excerpt.like(term)))

    if filters.get('category'):
        category = db.query(Category).filter(Category.slug == filters['category']).first()
        if category:
            q = q.filter(Post.categoryId == category.id)

    if filters.get('tag'):
        q = (
            q.join(PostTag, PostTag.postId == Post.id)
            .join(Tag, Tag.id == PostTag.tagId)
            .filter(Tag.slug == filters['tag'])
            .distinct()
        )

    total = q.count()
    posts = q.order_by(order_by_clause(Post, sort, order)).offset(offset).limit(limit).all()

    return {
        'items': [_to_public_post(p) for p in posts],
        'meta': {'page': page, 'limit': limit, 'total': total, 'totalPages': (total + limit - 1) // limit},
    }


def get_public_post_by_slug(db, slug):
    post = (
        db.query(Post)
        .filter(Post.slug == slug, Post.deletedAt.is_(None), *db_public_condition())
        .first()
    )
    if not post:
        raise AppError.not_found('Post not found')
    return _to_public_post(post)


def list_public_categories(db, filters):
    page, limit, offset, sort, order = parse_pagination(filters)

    q = db.query(Category).filter(Category.status == 'active')
    if filters.get('search'):
        q = q.filter(Category.name.like('%{}%'.format(filters['search'])))

    total = q.count()
    categories = q.order_by(order_by_clause(Category, sort, order)).offset(offset).limit(limit).all()

    return {
        'items': categories,
        'meta': {'page': page, 'limit': limit, 'total': total, 'totalPages': (total + limit - 1) // limit},
    }


def get_public_category_by_slug(db, slug):
    category = db.query(Category).filter(Category.slug == slug, Category.status == 'active').first()
    if not category:
        raise AppError.not_found('Category not found')
    return category


def list_public_tags(db, filters):
    page, limit, offset, sort, order = parse_pagination(filters)

    q = db.query(Tag)
    if filters.get('search'):
        q = q.filter(Tag.name.like('%{}%'.format(filters['search'])))

    total = q.count()
    tags = q.order_by(order_by_clause(Tag, sort, order)).offset(offset).limit(limit).all()

    return {
        'items': tags,
        'meta': {'page': page, 'limit': limit, 'total': total, 'totalPages': (total + limit - 1) // limit},
    }


def get_public_tag_by_slug(db, slug):
    tag = db.query(Tag).filter(Tag.slug == slug).first()
    if not tag:
        raise AppError.not_found('Tag not found')
    return tag


def search_public(db, query, filters):
    page, limit, offset = parse_pagination(filters)[:3]

    term = '%{}%'.format(query)
    q = (
        db.query(Post)
        .filter(
            Post.deletedAt.is_(None),
            Post.status == Const.POST_PUBLISHED,
            Post.visibility == Const.POST_VISIBILITY_PUBLIC,
            db.or_(Post.title.like(term), Post.excerpt.like(term), Post.content.like(term)),
        )
    )

    total = q.count()
    posts = q.order_by(Post.publishedAt.desc()).offset(offset).limit(limit).all()

    return {
        'items': [_to_public_post(p) for p in posts],
        'meta': {'page': page, 'limit': limit, 'total': total, 'totalPages': (total + limit - 1) // limit},
    }